hbase-server/src/main/java/org/apache/hadoop/hbase/security/access/SnapshotScannerHDFSAclHelper.java

```diff
@@ -0,0 +1,745 @@
+import org.apache.hbase.thirdparty.com.google.common.collect.Sets;
+import org.apache.hbase.thirdparty.com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+/**
+ * A helper to set HBase granted user access acl and default acl over hFiles.
+ */
+@InterfaceAudience.Private
+public class SnapshotScannerHDFSAclHelper implements Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(SnapshotScannerHDFSAclHelper.class);
+
+  public static final String USER_SCAN_SNAPSHOT_ENABLE = "hbase.user.scan.snapshot.enable";
+  public static final String USER_SCAN_SNAPSHOT_THREAD_NUMBER =
+      "hbase.user.scan.snapshot.thread.number";
+  // the tmp directory to restore snapshot, it can not be a sub directory of HBase root dir
+  public static final String SNAPSHOT_RESTORE_TMP_DIR = "hbase.snapshot.restore.tmp.dir";
+  public static final String SNAPSHOT_RESTORE_TMP_DIR_DEFAULT =
+      "/hbase/.tmpdir-to-restore-snapshot";
+  // If enable this feature, set public directories permission to 751
+  public static final String COMMON_DIRECTORY_PERMISSION =
+      "hbase.user.scan.snapshot.common.directory.permission";
+  public static final String COMMON_DIRECTORY_PERMISSION_DEFAULT = "751";
+  // If enable this feature, set restore directory permission to 703
+  public static final String SNAPSHOT_RESTORE_DIRECTORY_PERMISSION =
+      "hbase.user.scan.snapshot.restore.directory.permission";
+  public static final String SNAPSHOT_RESTORE_DIRECTORY_PERMISSION_DEFAULT = "703";
+
+  private Admin admin;
+  private final Configuration conf;
+  private FileSystem fs;
+  private PathHelper pathHelper;
+  private ExecutorService pool;
+
+  public SnapshotScannerHDFSAclHelper(Configuration configuration, Connection connection)
+      throws IOException {
+    this.conf = configuration;
+    this.pathHelper = new PathHelper(conf);
+    this.fs = pathHelper.getFileSystem();
+    this.pathHelper = new PathHelper(conf);
+    this.pool = Executors.newFixedThreadPool(conf.getInt(USER_SCAN_SNAPSHOT_THREAD_NUMBER, 10),
+      new ThreadFactoryBuilder().setNameFormat("hdfs-acl-thread-%d").setDaemon(true).build());
+    if (connection == null) {
+      connection = ConnectionFactory.createConnection(conf);
+    }
+    this.admin = connection.getAdmin();
+  }
+
+  @Override
+  public void close() {
+    if (pool != null) {
+      pool.shutdown();
+    }
+  }
+
+  public void setCommonDirPermission() throws IOException {
+    // Set public directory permission to 751 to make all users have access permission.
+    // And we also need the access permission of the parent of HBase root directory, but
+    // it's not set here, because the owner of HBase root directory may don't own permission
+    // to change it's parent permission to 751.
+    // The {root/.tmp} and {root/.tmp/data} directories are created to make global user HDFS
+    // acls can be inherited.
+    Path[] paths = new Path[] { pathHelper.getRootDir(), pathHelper.getDataDir(),
```

Comment:
One concern is : if someone add a path under the hbase.root.dir, then we must forget to grant the permission here... Also the dev won't  realize he/her is missing something like this.  I mean if we plan to depend on the dirs under hbase.root,  we need a UT to know if any newly added dirs or some removed dirs in future release, we must know it will break the snapshot ACL feature ...
I mean need a UT like : start the clusters,   assert the dirs under hbase.root,  if all expected, then pass, otherwise provide information to show dev say that we need to fix the snapshot ACL feature.